library/core/src/error.rs

```diff
@@ -219,13 +214,18 @@ impl<'a> dyn Error + 'a {
     /// Request a reference of type `T` as context about this error.
     #[unstable(feature = "error_generic_member_access", issue = "99301")]
     pub fn request_ref<T: ?Sized + 'static>(&'a self) -> Option<&'a T> {
-        core::any::request_ref(self)
+        request_ref(self)
     }
 
     /// Request a value of type `T` as context about this error.
     #[unstable(feature = "error_generic_member_access", issue = "99301")]
     pub fn request_value<T: 'static>(&'a self) -> Option<T> {
-        core::any::request_value(self)
+        request_value(self)
+    }
```

Comment:
Could you add these to the `Error` trait with a `where Self: Sized` bound, so they can also be used with normal error types that are not `dyn Error`. This is useful since there is no other way to create a `Request`.

Reply:
Do you mean we should keep the `dyn Error` impls and simply add new `Error` trait methods with default function bodies? Or should we discard the `dyn Error` impls? If `request_value` and `request_ref` are trait methods then they automatically become available for `dyn Error`s, right?

Reply:
Oh I think I may have realized the answer to this question -- we need both because the `where Self: Sized` bound doesn't apply to `dyn Error`, right?

Reply:
Okay I think my last comment is essentially correct because when I remove all `request_ref` and `request_value` from the variations on `dyn Error` I get
```
   Compiling std v0.0.0 (/home/wayne/projects/rust-lang/rust/library/std)
error: the `request_ref` method cannot be invoked on a trait object
   --> library/std/src/error.rs:460:53
    |
460 |         let backtrace = (&self.error as &dyn Error).request_ref();
    |                                                     ^^^^^^^^^^^
    |
   ::: /home/wayne/projects/rust-lang/rust/library/core/src/error.rs:218:21
    |
218 |         where Self: Sized
    |                     ----- this has a `Sized` requirement

error: the `request_ref` method cannot be invoked on a trait object
   --> library/std/src/error.rs:464:73
    |
464 |                 .map(|source| source.sources().find_map(|source| source.request_ref()))
    |                                                                         ^^^^^^^^^^^
    |
   ::: /home/wayne/projects/rust-lang/rust/library/core/src/error.rs:218:21
    |
218 |         where Self: Sized
    |                     ----- this has a `Sized` requirement

error: could not compile `std` (lib) due to 2 previous errors
```
Line 460 can be fixed by changing `(&self.error as &dyn Error)` to `self.error` but on line 464 the `source` in `source.request_ref()` is the return value of the `Error` trait method `source` which has type `&(dyn Error + 'static)`.

Just to be clear, we do want to support generic member access on `dyn Error`, right? But when I don't remove the `request_value` and request_ref` methods from the `dyn Error` variations I get this kind of error:

```
error[E0034]: multiple applicable items in scope
   --> library/core/src/error.rs:404:22
    |
404 |         <dyn Error>::request_value(self)
    |                      ^^^^^^^^^^^^^ multiple `request_value` found
    |
note: candidate #1 is defined in an impl for the type `(dyn error::Error + 'a)`
   --> library/core/src/error.rs:290:5
    |
290 |     pub fn request_value<T: 'static>(&'a self) -> Option<T> {
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
note: candidate #2 is defined in an impl for the type `(dyn error::Error + Send + 'static)`
   --> library/core/src/error.rs:368:5
    |
368 |     pub fn request_value<T: 'static>(&self) -> Option<T> {
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
note: candidate #3 is defined in an impl for the type `(dyn error::Error + Send + Sync + 'static)`
   --> library/core/src/error.rs:403:5
    |
403 |     pub fn request_value<T: 'static>(&self) -> Option<T> {
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
note: candidate #4 is defined in the trait `error::Error`
   --> library/core/src/error.rs:238:5
    |
238 | /     fn request_value<'a, T: 'static>(&'a self) -> Option<T>
239 | |         where Self: Sized
    | |_________________________^
help: disambiguate the method for candidate #4
    |
404 |         <dyn error::Error as error::Error>::request_value(self)
    |         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

For line 404 if I try doing something like `self.request_value()`, I get the aforementioned `error: the `request_value` method cannot be invoked on a trait object` error. Definitely open to suggestions here...

Reply:
Ah, I didn't think they would conflict thanks to the `Self: Sized` bound, but that's probably not possible due to the way name resolution works. In that case it's probably better to keep them as free functions outside the trait, and remove both the ones in the trait (that I mistakenly suggested) and the ones in `impl dyn Error`. The goal here is to have one consistent way to request values that works for both concrete error types and `dyn Error`.

Reply:
Oh interesting, so don't even keep the `impl dyn Error` methods? Will do!

Reply:
Okay so I tried using free functions and removing the `impl dyn Error` methods as suggested but I am now running into this error:
```
error[E0277]: the size for values of type `dyn core::error::Error` cannot be known at compilation time
   --> library/std/src/error.rs:466:78
    |
466 |                 .map(|source| source.sources().find_map(|source| request_ref(source)))
    |                                                                  ----------- ^^^^^^ doesn't have a size known at compile-time
    |                                                                  |
    |                                                                  required by a bound introduced by this call
    |
    = help: the trait `core::marker::Sized` is not implemented for `dyn core::error::Error`
note: required by a bound in `request_ref`
   --> /home/wayne/projects/rust-lang/rust/library/core/src/error.rs:409:36
    |
409 | pub fn request_ref<'a, T>(err: &'a impl Error) -> Option<&'a T>
    |                                    ^^^^^^^^^^ required by this bound in `request_ref`
```

This after removing the `+ &Sized` bound in that parameter's type:
```
-fn request_value<'a, T>(err: &'a (impl Error + ?Sized)) -> Option<T>
+pub fn request_value<'a, T>(err: &'a impl Error) -> Option<T>
```

The error is literally saying that `impl Error` implies a `core::marker::Sized` bound on the type parameter, but I don't see that mentioned in [the docs](https://doc.rust-lang.org/reference/types/impl-trait.html) on `impl Error` which I understand to be telling me that in parameter position `impl Error` is similar in purpose to a generic type parameter.

My second or third thought about what might be going on here is that there could be a `Self: Sized` bound on a different `Error` trait method, but I don't see that. But then I thought that maybe there is an `impl Error` block somewhere doing that and sure enough I found it: https://github.com/rust-lang/rust/blob/master/library/core/src/error.rs#L437

Could this line be the reason that `impl Error` implies a `Sized` bound?

Another question that's been stewing in my puny brain over the past couple days while pondering this is, what does this mean?:

> The goal here is to have one consistent way to request values that works for both concrete error types and dyn Error

In my mind the prototypical use case of generic member access for `Error` (or for any type for that matter) is that we want to offer the ability for one library that knows nothing about the concrete `Error` types it might be dealing with to access a value that might potentially be provided by some concrete types but not others. To be more specific, I would like to see a "frontend" error-handling library like `anyhow` be able to access data (specifically, backtraces) stored on errors it can't be certain is available because the concrete type of an `Error` that such frontends encounter will never be explicitly known by `anyhow` (my understanding is that `anyhow` would be usually be accessing `dyn Error`).

I'll admit that what I just wrote confuses me a little, but the train of reasoning I'm trying to reach here is -- does it make sense to want to perform generic member access on concrete error types? If you have access to the concrete type, you could just perform field access of the member that you're interested in. But I suspect I am misunderstanding what you meant (hence the word salad here).

Instead of the free function, would it make sense to instead have `request_ref` and `request_value` methods defined on both `dyn Error` and  the aforementioned `impl<'a, T: Error + ?Sized> Error for &'a T`? Kind of getting into "don't know what i'm talking about" territory here in the sense that I'm not sure if this addresses your concern about consistency...

Reply:
`T: Sized` bounds are always implicitly added unless specifically opted-out-of with `T: ?Sized`. This has nothing to do with `Error` but is instead inherent to the Rust language.

You shouldn't need to change the signature of `request_value`/`request_ref`, it is already correct with `err: &'a (impl Error + ?Sized)`.

Reply:
d'oh! Thanks for the quick response.

Reply:
I think the reason I removed that was because I was confused about some other compiler errors...gonna take a closer at those now.

Reply:
Okay this is the error that may have been confusing me before:
```
error[E0277]: the size for values of type `dyn std::error::Error` cannot be known at compilation time
   --> library/core/tests/any.rs:199:36
    |
199 |     assert_eq!(request_value::<u8>(&obj), None);
    |                ------------------- ^^^^ doesn't have a size known at compile-time
    |                |
    |                required by a bound introduced by this call
    |
    = help: the trait `Sized` is not implemented for `dyn std::error::Error`
    = help: the trait `std::error::Error` is implemented for `Box<T>`
    = note: required for `Box<dyn std::error::Error>` to implement `std::error::Error`
note: required by a bound in `request_value`
   --> /home/wayne/projects/rust-lang/rust/library/core/src/error.rs:387:44
    |
387 | pub fn request_value<'a, T>(err: &'a (impl Error + ?Sized)) -> Option<T>
    |                                            ^^^^^ required by this bound in `request_value`
```

Since I didn't know that `?` meant to opt-out of `T: Sized` bounds I incorrectly thought removing `?Sized` would fix this, but it actually just makes the compile fail in a slightly different way.

The full body of the specific test that's failing is:
```
// Test the provide and request mechanisms with a boxed trait object.
#[test]
fn test_provider_boxed() {
    let obj: Box<dyn std::error::Error> =
        Box::new(SomeConcreteType { some_string: "hello".to_owned() });

    assert_eq!(request_value::<String>(&*obj).unwrap(), "bye");
    assert_eq!(request_value::<u8>(&obj), None);
}
```

The definition of `request_value` used here is:
```
pub fn request_value<'a, T>(err: &'a (impl Error + ?Sized)) -> Option<T>
where
    T: 'static,
{
    request_by_type_tag::<'a, tags::Value<T>>(err)
}
```

I'm still a confused by this, but I'm also getting super tired so I'll try to come back to it in the morning.

Reply:
It seems that `Box<E>` only implement `Error` when `E: Error + Sized`. See #104485.

You can work around it by using `&*obj` in the test.